### ISSUE TITLE
Serve Vector Tiles from Valhalla Data

### ISSUE THREAD

**Issue by kevinkreiser** [valhalla/valhalla#1703]
We should have more tools for visually inspecting the valhalla data and ways to visually understand how sif/odin interact with it.

We currently have a tool that lets us see most of this data in json form however, reading enourmous amounts of properties in plain text is quite tedious and error prone.

What we propose is a vector tile server that will generate [mapbox vector tile spec encoded](https://github.com/mapbox/vector-tile-spec/tree/master/2.1) tiles. These are the defacto-standard when it comes to visualizing osm and derived datasets such as valhalla. This should allow for the creation of a visualization tool and accompanying styles to aide in debugging or just understanding how libvalhalla makes the decisions that it does when routing.

**Comment by Guo-astro** [comment]
@kevinkreiser  Hi! Really interested in this, where is a good start point?

**Comment by kevinkreiser** [comment]
i have an unpushed branch that has the request logic for this but none of the tile cutting logic yet (just the pbf definitions). i reckon i should write a more detailed plan here so that others can take up the work. give me a bit and ping back if i dont get back to it shortly!

**Comment by ianthetechie** [comment]
@kevinkreiser I'd be curious at having a look into implementing this. As I was building the LSV profile (and unrelatedly investigating random complaints over the years haha), I've long wished for something like this to get better visibility into how the graph looks in specific areas / why paths weren't taken.

(The expansion endpoint is useful, but in very different ways.)

**Comment by kevinkreiser** [comment]
yeah the locate endpoint also gives a slice of this data but not in the same slippy map way that we are used to. to me, serializing all the properties in mvt would be the best bet for doing this. there are some serious problems though:

1. at high zoom numbers everything is fine: its easy to create the geoms and properties for individual edges and everyone is happy
2. at low zooms life is hard: they start to become intractable around z12 and under or so and the only remedy is, as we do in the graph algorithm, cull the search space 
3. so we need to have a configuration to specify how that culling happens there are a few things to do:
  1. drop the least importance properties
  2. merge features with matching properties as much as possible
  3. drop features that are too small to be visible
  4. generalize the geometry of the remaining features appropriate to the target zoom level
4. all of that is just to make it usable on the client side (ie not too much data) we still have the problem that it will take a while to compute the tiles
5. so we need to add caching to disk and probably an optional pre-rendering (seeding) step that makes the low zoom tiles before we start a service (could build them when we build the routing tiles)
6. for a first pass i would concentrate on rendering only high zoom tiles where we dont need to do any dropping/merging/generalizing

a couple other things.. we should use vtzero for serialization and we should probably also add point features for graph nodes at the high zooms. when i originally wrote this i was thinking about a visual layer to go with the routing but not from the standpoint of inspection really just from the standpoint of usability. if we had this feature it would be quite easy to recreate a little "garmin"-like application with simplistic display and excellent routing.

**Comment by ianthetechie** [comment]
Thanks for the writeup, Kevin! Yeah, more or less what I was thinking.

Agreed: the first phase should concentrate only on high zoom levels since that's "easy mode"; no need to worry about removing features or generalizing. Selecting what goes into the tiles wider than say z13 or so, generalizing complex geometries, etc. are phase 2 / 3 items IMO.

It's kind of funny how we're both approaching it from different directions. My _first_  thought was re: the debugging/inspection/visibility for server operators, devs, and (aspirationally) the OSM community. But I also agree there's huge potential for visualization (ex: which roads of a particular class, access, etc. are available at a zoom level). Hence your third point.

Dreaming for a moment here, it would actually be nice for the user to be able to specify some of this stuff up front (ex: what's the view of a cyclist or an LSV driver) as a primitive method of filtering. Valhalla isn't going to be the next planet tile server, _but_ it has a very unique space in giving people a view of the world from the perspective of routers, and I think that's valuable.

I hadn't given a ton of thought to use of Valhalla directly at low zoom TBH. I assumed that the slow gen speed would be fine and anyone interested in such tiles would be either 1) OK with the long render times, or 2) would have a pregen solution of their own given their intended use case (ex: MBTiles or PMTiles for fairly static workloads, Varnish for more dynamic ones, or a combination).

Agreed re: vtzero and point features for nodes at high zoom (14? some are saying 15 for this stuff these days...).

**Comment by kevinkreiser** [comment]
> Dreaming for a moment here, it would actually be nice for the user to be able to specify some of this stuff up front (ex: what's the view of a cyclist or an LSV driver) as a primitive method of filtering.

imho that is a job for the style with which you render the data, its made for that purpose. this also allows us to avoid the annoyance of cache busting by changing the filtering parameters. imho let the style do the filtering

**Comment by kevinkreiser** [comment]
> I hadn't given a ton of thought to use of Valhalla directly at low zoom TBH. I assumed that the slow gen speed would be fine and anyone interested in such tiles would be either 1) OK with the long render times, or 2) would have a pregen solution of their own given their intended use case (ex: MBTiles or PMTiles for fairly static workloads, Varnish for more dynamic ones, or a combination). 

i agree with `1` but for number `2` i think a simple solution could involve a some straight forward config parameters like:

```json
...
"visual_tiles_cache_dir": "some_dir",
"visual_tiles_cache_zoom_cutoff": 10,
...
```

then the code can just check the directory to see if the mvt is there before rendering it. and if we just rendered a fresh tile AND its zoom is the cutoff or less we throw it into the directory as well.

**Comment by ianthetechie** [comment]
> imho that is a job for the style with which you render the data, its made for that purpose. this also allows us to avoid the annoyance of cache busting by changing the filtering parameters. imho let the style do the filtering

My thinking behind having some filtering in Valhalla is that the client doesn't necessarily know what access restrictions go into a profile. Being able to get a view of the routable world _as a golf cart or a pedestrian (for example)_ seems useful to me, and I think offering optional filtering is a reasonable way to do it.

Or is there a way I'm missing that we could easily generate profile-specific styles? (Not sure I like this idea because then it assumes the user wants data styled a certain way, and it seems like it'd be a pain to ensure accuracy of such filters since the source of truth is a predicate function in C++, not some declarative rules which can be transformed into both access code and style rules.)

Agreed though re: it turning into an annoying cache busting problem :P Hence my first thought was to make caching a "you bring your own solution" kind of thing 😅

**Comment by ianthetechie** [comment]
> id really rather not add yet another dependency to valhalla if we can think of simple/easy enough ways to solve the problem with out them. dont get me wrong pmtiles and mbtiles are fine (mbtiles is super easy because we already have sqlite as a dependency) and pmtiles is also great because it uses all the same tricks we use in our own data format but still where we can avoid complication at least to start i think its a good idea. just my 2 cents though!

Agreed :) I'm thinking extremely simple and no extra deps besides absolutely necessary (ex: vtzero).

**Comment by ianthetechie** [comment]
I finally had a bit more of a need for this at the end of last year when debugging some graph issues, so I hacked up a tool.

The "bad" news is that I didn't really have the mental energy to figure out how to do it in C++ with vtzero etc. from scratch. Instead, I went for a lower effort path using Tippecanoe. Tippecanoe obviously can't understand Valhalla data directly, so what I settled on was a GeoJSON dump of all the edges in the routing graph; then I let Tippecanoe figure out how to tile it 😂

The obvious question is probably "isn't that super resource intensive." Surprisingly, it's not too bad. You can do the whole planet in a few hours on modest hardware. And to avoid eating up a positively obscene amount of disk space, I zstd compressed the GeoJSON files when writing them out. Then pipe em into tippecanoe, which does a pretty good job sorting everything into far more efficient data structures.

Here's how to do it + the GeoJSON exporter code in case anyone wants to do something similar: https://github.com/stadiamaps/valinor/tree/main/valinor-export-graph.

**Comment by chrstnbwnkl** [comment]
For those really just interested in viewing edges and their properties for debugging purposes, I have written a little tool that dumps tiles to flatgeobuffers (a vector format with a spatial index for fast rendering). It is supported by most GIS software (I use QGIS). I use GDAL, which is super fast, so exporting whole Europe should take less than 5 minutes with decent hardware :)  https://github.com/chrstnbwnkl/valhalla-tools happy to review contributions for extending the list of supported attributes!

**Comment by ianthetechie** [comment]
Nice; thanks for sharing @chrstnbwnkl! I should look at teh current state of GDAL + tippecanoe again and see if there's a more efficient middle path than GeoJSON. That's some impressive timing on the export!

**Comment by SiarheiFedartsou** [comment]
Hey, I used OSRM a lot during recent years, they have this feature and they actually do quite simple things:
- they don't render anything for zoom level <12 (I can be wrong with exact number) 
- they render it on the fly

So I quickly tried to hack something using their code as inspiration and got this:

https://github.com/user-attachments/assets/3db884f2-0ba8-4d1f-bd8f-319860d95311

I don't want to show code yet - it is too dirty yet to be even a draft PR 😄 

IMO it works pretty good and I'd like to finish it... 

Any opinions? Should I continue?

**Comment by ianthetechie** [comment]
Tallinn 🇪🇪😍

Idk if I can give you a good answer without knowing the goals / tradeoffs of your solution haha. It seems like we already have two possible solutions published with slightly different decisions. I can't see how a third would make the world worse! But maybe if you describe what your goals were, how it works, etc. you'd get better feedback. 

(As an aside, one thing which I probably will end up building at some point that makes very different tradeoffs is a live tile renderer with a server API. This would mean every tile access is slower and uses some resources. But it might be more convenient for many uses than having to pregen everything.)

**Comment by SiarheiFedartsou** [comment]
> It seems like we already have two possible solutions published with slightly different decisions


Wait, what existing solutions do you mean? 😀


My goal is simple: I want to be able to have more insights on why routing engine made certain decisions. Also what would be really nice is to be able to dynamically see live traffic speeds applied to traffic.tar... WDYT?

**Comment by SiarheiFedartsou** [comment]
> how it works


Ah, basically I added tile(x, y, z) method to actor_t which renders mapbox vector tiles on the fly using vtzero. Ie what you see on my video is not just rendered GeoJSON or something like this, but trivial HTML with map + simple tile server which calls Valhalla to render tiles...

**Comment by nilsnolde** [comment]
Can't play the video, but yeah, I'd (still) highly appreciate a MVT endpoint.

**Comment by SiarheiFedartsou** [comment]
> Can't play the video

Hm, why? Something wrong with format of it?

**Comment by nilsnolde** [comment]
works on firefox:) (I actually use brave, basically chromium)

**Comment by SiarheiFedartsou** [comment]
Please see code here https://github.com/valhalla/valhalla/pull/5663/files

I want to discuss the approach first before I'll invest time into finishing it...

**Comment by SiarheiFedartsou** [comment]
https://github.com/user-attachments/assets/9c514273-7ac6-4d0c-ac78-5f8b3b78eff2

And yet another video: added nodes + label with `DirectedEdge::speed` - just a demo of how we could show speed data right on the graph (`DirectedEdge::speed` is probably not that useful as something more dynamic like data from traffic tiles).
Also started dynamically showing different roads based on road class and zoom - just looks a bit better and doesn't impact performance it seems.
